File: LeetCode/Kotlin/s1438.kt
fun main() {

    val sl = Solution1438();

    val nums = intArrayOf(10,1,2,4,7,2)
    val  limit = 5

    println(sl.longestSubarray(nums, limit))
}

class Solution1438 {
    fun longestSubarray(nums: IntArray, limit: Int): Int =
        nums.fold(Triple(mutableListOf<Int>(), mutableListOf<Int>(), 0)) { (maxD, minD, i), n ->
            while (maxD.isNotEmpty() && n > maxD.last()) maxD.removeLast()
            while (minD.isNotEmpty() && n < minD.last()) minD.removeLast()
            maxD.add(n)
            minD.add(n)
            (if (maxD.first() - minD.first() > limit) {
                if (maxD.first() == nums[i]) maxD.removeFirst()
                if (minD.first() == nums[i]) minD.removeFirst()
                1 } else 0)
                .let { Triple(maxD, minD, i + it)}
        }.let { nums.size - it.third }
}

File: LeetCode/Rust/s1438.rs
struct Solution1438 {}

impl Solution1438 {
    pub fn longest_subarray(nums: Vec<i32>, limit: i32) -> i32 {
        use std::collections::VecDeque;
        let mut max_d: VecDeque<i32> = VecDeque::new();
        let mut min_d: VecDeque<i32> = VecDeque::new();
        let mut i = 0;
        for n in nums.iter() {
            while max_d.back().is_some_and(|x| *n > *x) { max_d.pop_back(); }
            while min_d.back().is_some_and(|x| *n < *x) { min_d.pop_back(); }
            max_d.push_back(*n);
            min_d.push_back(*n);
            if max_d[0] - min_d[0] > limit {
                if max_d[0] == nums[i] { max_d.pop_front(); }
                if min_d[0] == nums[i] { min_d.pop_front(); }
                i += 1;
            }
        }
        (nums.len() - i) as i32
    }
}
